solidity/contracts/isms/multisig/AbstractMultisigIsm.sol

```diff
@@ -54,53 +54,60 @@ abstract contract AbstractMultisigIsm is IMultisigIsm {
         view
         returns (bool)
     {
-        require(_verifyMerkleProof(_metadata, _message), "!merkle");
-        require(_verifyValidatorSignatures(_metadata, _message), "!sigs");
-        return true;
+        return verifyValidatorSignatures(_metadata, _message);
     }
 
-    // ============ Internal Functions ============
+    function computeDigest(bytes calldata _metadata, bytes calldata _message)
+        public
+        pure
+        returns (bytes32 digest)
+    {
+        bytes32 root;
+        bytes32 signedMessageId;
 
-    /**
-     * @notice Verifies the merkle proof of `_message` against the provided
-     * checkpoint.
-     * @param _metadata ABI encoded module metadata (see MultisigIsmMetadata.sol)
-     * @param _message Formatted Hyperlane message (see Message.sol).
-     */
-    function _verifyMerkleProof(
-        bytes calldata _metadata,
-        bytes calldata _message
-    ) internal pure returns (bool) {
-        // calculate the expected root based on the proof
-        bytes32 _calculatedRoot = MerkleLib.branchRoot(
-            Message.id(_message),
-            MultisigIsmMetadata.proof(_metadata),
-            Message.nonce(_message)
+        if (
+            MultisigIsmMetadata.merkleType(_metadata) ==
+            MultisigIsmMetadata.MerkleType.ROOT
+        ) {
+            signedMessageId = Message.id(_message);
+            root = MultisigIsmMetadata.root(_metadata);
+        } else {
+            bytes32[32] memory proof;
+            (signedMessageId, proof) = MultisigIsmMetadata.idAndProof(
+                _metadata
+            );
+            root = MerkleLib.branchRoot(
+                signedMessageId,
```

Comment:
This needs to be `Message.id(_message)` or else the ISM will verify arbitrary messages 😬

Reply:
Please add a test case that would have caught this